File: labodc-web-portal/src/services/auth.service.ts
// Authentication Service
import axiosInstance from './api/axios.config';
import { IApiResponse } from '@/types/api.types';
import {
  ILoginRequest,
  IRegisterRequest,
  IAuthResponse,
  IForgotPasswordRequest,
  IResetPasswordRequest,
} from '@/types/auth.types';

const AUTH_ENDPOINTS = {
  LOGIN: '/auth/login',
  REGISTER: '/auth/register',
  LOGOUT: '/auth/logout',
  REFRESH: '/auth/refresh',
  FORGOT_PASSWORD: '/auth/forgot-password',
  RESET_PASSWORD: '/auth/reset-password',
  VERIFY_EMAIL: '/auth/verify-email',
};

/** Backend wraps auth in ApiResponse: { success, message, data: IAuthResponse } */
export const authService = {
  /**
   * Login - returns full API body { success, message, data }; slice uses .data
   */
  login: async (data: ILoginRequest): Promise<IApiResponse<IAuthResponse>> => {
    const response = await axiosInstance.post<IApiResponse<IAuthResponse>>(AUTH_ENDPOINTS.LOGIN, data);
    return response.data;
  },

  /**
   * Register - backend returns AuthResponse in .data; we then login and return login body
   */
  register: async (data: IRegisterRequest): Promise<IApiResponse<IAuthResponse>> => {
    const regRes = await axiosInstance.post<IApiResponse<IAuthResponse>>(AUTH_ENDPOINTS.REGISTER, {
      email: data.email,
      password: data.password,
      confirmPassword: data.confirmPassword,
      role: data.role,
    });
    // Backend register already returns token in regRes.data.data; we can use it or re-login
    if (regRes.data?.data?.token) {
      return regRes.data;
    }
    const loginResponse = await axiosInstance.post<IApiResponse<IAuthResponse>>(AUTH_ENDPOINTS.LOGIN, {
      email: data.email,
      password: data.password,
    });
    return loginResponse.data;
  },

  /**
   * Logout
   */
  logout: async (): Promise<void> => {
    await axiosInstance.post(AUTH_ENDPOINTS.LOGOUT);
  },

  /**
   * Refresh token
   */
  refreshToken: async (refreshToken: string): Promise<IAuthResponse> => {
    const response = await axiosInstance.post<IAuthResponse>(AUTH_ENDPOINTS.REFRESH, {
      refreshToken,
    });
    return response.data;
  },

  /**
   * Forgot password
   */
  forgotPassword: async (data: IForgotPasswordRequest): Promise<void> => {
    await axiosInstance.post(AUTH_ENDPOINTS.FORGOT_PASSWORD, data);
  },

  /**
   * Reset password
   */
  resetPassword: async (data: IResetPasswordRequest): Promise<void> => {
    await axiosInstance.post(AUTH_ENDPOINTS.RESET_PASSWORD, data);
  },

  /**
   * Verify email
   */
  verifyEmail: async (token: string): Promise<void> => {
    await axiosInstance.post(`${AUTH_ENDPOINTS.VERIFY_EMAIL}/${token}`);
  },
};

File: labodc_mobile/lib/services/auth_service.dart
import 'package:labodc_mobile/core/constants/api_endpoints.dart';
import 'package:labodc_mobile/models/auth_model.dart';
import 'package:labodc_mobile/models/api_response.dart';
import 'package:labodc_mobile/services/api_service.dart';
import 'package:labodc_mobile/services/storage_service.dart';

/// Authentication Service
class AuthService {
  final _api = ApiService();
  final _storage = StorageService();
  
  /// Login user
  Future<LoginResponse> login(String email, String password) async {
    try {
      final request = LoginRequest(email: email, password: password);
      final response = await _api.post(
        ApiEndpoints.login,
        data: request.toJson(),
      );
      
      final loginResponse = LoginResponse.fromJson(response.data);
      
      if (loginResponse.success && loginResponse.data != null) {
        final authData = loginResponse.data!;
        
        // Save tokens (match web portal storage keys)
        await _storage.saveAccessToken(authData.token);
        if (authData.refreshToken != null) {
          await _storage.saveRefreshToken(authData.refreshToken!);
        }
        
        // Save user info
        await _storage.saveUserData({
          'userId': authData.userId,
          'email': authData.email,
          'role': authData.role,
          'status': authData.status,
          'emailVerified': authData.emailVerified,
        });
        await _storage.saveUserRole(authData.role);
      }
      
      return loginResponse;
    } catch (e) {
      rethrow;
    }
  }
  
  /// Register new user
  Future<ApiResponse> register(RegisterRequest request) async {
    try {
      final response = await _api.post(
        ApiEndpoints.register,
        data: request.toJson(),
      );
      
      return ApiResponse.fromJson(response.data, null);
    } catch (e) {
      rethrow;
    }
  }
  
  /// Forgot password
  Future<ApiResponse> forgotPassword(String email) async {
    try {
      final response = await _api.post(
        ApiEndpoints.forgotPassword,
        data: {'email': email},
      );
      
      return ApiResponse.fromJson(response.data, null);
    } catch (e) {
      rethrow;
    }
  }
  
  /// Reset password
  Future<ApiResponse> resetPassword({
    required String token,
    required String otp,
    required String newPassword,
  }) async {
    try {
      final response = await _api.post(
        ApiEndpoints.resetPassword,
        data: {
          'token': token,
          'otp': otp,
          'newPassword': newPassword,
        },
      );
      
      return ApiResponse.fromJson(response.data, null);
    } catch (e) {
      rethrow;
    }
  }
  
  /// Verify email
  Future<ApiResponse> verifyEmail(String token) async {
    try {
      final response = await _api.post(
        ApiEndpoints.verifyEmail,
        data: {'token': token},
      );
      
      return ApiResponse.fromJson(response.data, null);
    } catch (e) {
      rethrow;
    }
  }
  
  /// Logout
  Future<void> logout() async {
    try {
      await _api.post(ApiEndpoints.logout);
    } catch (e) {
      // Continue with local logout even if API fails
    } finally {
      await _storage.clear();
    }
  }
  
  /// Check if user is logged in
  Future<bool> isLoggedIn() async {
    final token = await _storage.getAccessToken();
    return token != null && token.isNotEmpty;
  }
  
  /// Get current user data
  Map<String, dynamic>? getCurrentUser() {
    return _storage.getUserData();
  }
  
  /// Get current user role
  String? getCurrentUserRole() {
    return _storage.getUserRole();
  }
}
